### ISSUE TITLE
Memory wise / chunked file upload

### ISSUE THREAD

**Issue by mikehouse** [nanomsg/nng#1480]
**Describe the bug**
Out of memory error when uploading large files.

**Expected behavior**
A file of any size can be uploaded.

**Actual Behavior**
Out of memory error.

**To Reproduce**
Try create the http handler using `nng_http_handler_alloc_file` function for some large file (size should be more than RAM available) and serve it to some http download request.

** Environment Details **
 - 1.5.1
 - All

Hi. I found this in the source code inside `static void http_handle_file(nni_aio *aio)` function

```c
	// This is a very simplistic file server, suitable only for small
	// files.  In the future we can use an AIO based file read, where
	// we read files a bit at a time, or even mmap them, and serve
	// them up chunkwise.  Applications could even come up with their own
	// caching version of the http handler.
```

I've tried different `aio` based ways to read file and write then to response, but it didn't work. Is there any example where I can look at to learn how to write data to `nni_http_res` object by buffer using `nni_aio` object ? I'd like to make a fork of your library to modify that function (`nng_http_handler_alloc_file`) to make it work for large files, but I cannot find a working solution for this.

I've tried something like this and other ways with no luck

```c
        nni_http_conn *   conn= nni_aio_get_input(aio, 2);
	size_t buf_size = 8192;
	uint8_t buffer[buf_size];
	
	FILE *file = fopen(hf->path, "r");
	size_t read;
	nng_iov iov;
	nng_aio *aio2 = NULL;

	nng_aio_alloc(&aio2, NULL, NULL);
//	nng_stream *stream = nng_aio_get_output(aio, 0);
	nng_stream *stream = nng_http_conn_get_stream(conn);
	while ((read = fread(buffer, 1, buf_size, file)) > 0) {
		iov.iov_len = read;
		iov.iov_buf = buffer;
		nng_aio_set_iov(aio2, 1, &iov);
		nng_stream_send(stream, aio2);
		nni_aio_wait(aio2);
	}

	fclose(file);

        nng_aio_stop(aio2);
	nng_aio_free(aio2);

	nni_aio_set_output(aio, 0, res);
	nni_aio_finish(aio, 0, 0);
```

I'm not c/c++ developer and some things look unfamiliar to me and hard to follow. This is one left unsolved problem and I cannot figure out how to approach to it. The link to the function https://github.com/nanomsg/nng/blob/55fb7f14dd95b81ca84e463749d1707da2f07c83/src/supplemental/http/http_server.c#L1393

I will appreciate any help.
Thank you very much for your work!

**Comment by gdamore** [comment]
First, I assume you want to *download*, not upload.  (Meaning NNG is the server, and is sending the file an HTTP user agent client.)

You'll have to replace http_handle_file entirely to provide streaming file support, and you'll need to implement a callback based scheme to do this properly.

I don't think you can just get the stream and send to it, because there is important work relating to ensuring that the headers are already sent out properly, and you need to send the ultimate file size apriori in a Content-Length header first.

All of this sounds somewhat complicated, and it is, which is why I opted not to do it initially.  But maybe I'll find some time to fix it properly so that this works regardless of file size.

**Comment by mikehouse** [comment]
> First, I assume you want to download, not upload. (Meaning NNG is the server, and is sending the file an HTTP user agent client.)

Yes, I mixed it up.

Thank you for such quick feedback on the issue.

> I don't think you can just get the stream and send to it, because there is important work relating to ensuring that the headers are already sent out properly, and you need to send the ultimate file size apriori in a Content-Length header first.
All of this sounds somewhat complicated, and it is ...

Yes, sounds almost impossible to me to implement these all myself )

> But maybe I'll find some time to fix it properly so that this works regardless of file size.

 Thank you, definitely will follow the project changes.

**Comment by colesnicov** [comment]
You might need a structure to send. Something like:

```
struct neco_s{
    size_t total;
    size_t actual;
    uint8_t data[buf_size];
}
```

use it some like this:

```
neco_s my_data;

fseek(file, 0L, SEEK_END); 
my_data.total = ftell(file);
fseek(file, 0, SEEK_SET);

while ((my_data.actual = fread(my_data.data, 1, buf_size, file)) > 0) { 
		iov.iov_len = sizeof(my_data);
		iov.iov_buf = my_data;
// or move data:
//              memmove(iov.iov_buf, my_data, sizeof(my_data));

		nng_aio_set_iov(aio2, 1, &iov);
		nng_stream_send(stream, aio2);
		nni_aio_wait(aio2);

               memset(&my_data, 0, sizeof(my_data));
	}

```

**Yes, there are errors in the code, it's just pseudo code, but the idea is probably clear.**

-------------------------
P.S.

*Why doesn't the author close this issue?*

**Comment by ZenLiuCN** [comment]
A mini demo of Chunked download.
``` c

#define NNG_IPC "http://127.0.0.1:8081"
#define NNG_CHUNK 256
#include <log_.h>
#include <nng/nng.h>
#include <nng/protocol/reqrep0/rep.h>
#include <nng/protocol/reqrep0/req.h>
#include <nng/supplemental/http/http.h>
#include <nng/supplemental/util/platform.h>
#include <sds_.h>

#define CHECK_RC(RC)                                                                                                   \
    if ((rc = RC) != 0) {                                                                                              \
        log_error("%s fail: %s", #RC, nng_strerror(rc));                                                               \
        return rc;                                                                                                     \
    }
#define CHECK_RC_ONLY(RC)                                                                                              \
    if ((rc = RC) != 0) {                                                                                              \
        log_error("%s fail: %s", #RC, nng_strerror(rc));                                                               \
    }
typedef struct {
    uint32_t       count;
    uint32_t       rct;
    nng_aio*       ca;
    nng_aio*       ra;
    nng_http_conn* conn;
    nng_http_req*  req;
    sds            chdr;
} ctx_t, *ctx;
static ctx_t cx = {};
static void  nng_http_server_defer_cb(nng_aio* aio, void* arg, int err) {
    log_error("cancel call fail %s", nng_strerror(err));
}
static void nng_http_server_cb(void* arg) {
    ctx c = arg;
    if (!c->chdr) c->chdr = sdsempty();
    else sdsclear(c->chdr);
start:
    switch (cx.count) {
    case 8: {
        c->count = -1;
        c->chdr  = sdscat(c->chdr, "0\r\n\r\n");
        nng_iov iov;
        iov.iov_buf = c->chdr;
        iov.iov_len = 5;
        nng_aio_set_iov(c->ca, 1, &iov);
        nng_http_conn_write(c->conn, c->ca);
        log_info("last chunk send");
        return;
    }
    case -1: {
        c->count = 0;
        log_info("end transport: %d", nng_aio_count(c->ca));
        nng_aio_reap(c->ca); 
        c->ca = NULL;
        nng_aio_set_output(c->ra, 0, NULL);
        nng_aio_finish(c->ra, 0);
        log_info("end transport for: <%lX>", c->ra);
        return;
    }
    default: {
        int rv;
        if (c->count == 0) {
            rv = nng_aio_result(c->ca);
            if (rv != 0) {
                log_error("last call fail %s", nng_strerror(rv));
                cx.count = -1;
                goto start;
            }
        } else {
            rv = nng_aio_result(c->ca);
            if (rv != 0) {
                log_error("last call fail %s", nng_strerror(rv));
                cx.count = -1;
                goto start;
            }
        }
        log_info("transport: %d for %d", nng_aio_count(c->ca), c->count);
        c->count++;
        c->chdr = sdscatprintf(c->chdr, "%x\r\n", 8);
        nng_iov iov[3];
        iov[0].iov_buf = c->chdr;
        iov[0].iov_len = sdslen(c->chdr);
        iov[1].iov_buf = (void*)"segment\n";
        iov[1].iov_len = 8;
        iov[2].iov_buf = "\r\n";
        iov[2].iov_len = 2;
        nng_aio_set_iov(c->ca, 3, &iov[0]);
        nng_http_conn_write(c->conn, c->ca);
        log_info("send chunk %d", c->count);
        return;
    }
    }
}
static void nng_http_server_handler(nng_aio* aio) {
    switch (cx.count) {
    case 0: { //** 8 count, 2 TaskThd for 1 conn
        cx.rct++;
        cx.req  = nng_aio_get_input(aio, 0);
        cx.conn = nng_aio_get_input(aio, 2);
        cx.ra   = aio;
        log_info("<HANLDER:0x%lX> <CONN:0x%lX> %d", nng_aio_get_input(aio, 1), cx.conn, cx.rct);
        nng_aio_defer(aio, nng_http_server_defer_cb, NULL);
        nng_aio_set_timeout(aio, 1000);
        //** send header
        nng_http_res* res = NULL;
        nng_http_res_alloc(&res);
        int rv = nng_http_res_set_header(res, "Transfer-Encoding", "Chunked");
        rv     = nng_http_res_set_status(res, 200);
        nng_aio_alloc(&cx.ca, nng_http_server_cb, &cx);
        nng_http_conn_write_res(cx.conn, res, cx.ca);
        return;
    }
    default: {
        log_error("not support parall: %d", cx.count);
    }
    }
}
static void hold_loop(void* arg) {
    for (;;) {
        SLEEP(500);
    }
}
int main(int argc, char* argv[]) {
    int               rc;
    nng_thread*       hold;
    nng_http_server*  server;
    nng_http_handler* handler;
    nng_url*          url;

    nng_init_set_parameter(NNG_INIT_NUM_TASK_THREADS, 2);
    nng_init_set_parameter(NNG_INIT_MAX_TASK_THREADS, 2);
    CHECK_RC(nng_thread_create(&hold, hold_loop, NULL));
    CHECK_RC(nng_url_parse(&url, "http://127.0.0.1:8081"));
    CHECK_RC(nng_http_server_hold(&server, url));
    CHECK_RC(nng_http_handler_alloc(&handler, "/", nng_http_server_handler));
    CHECK_RC(nng_http_handler_set_tree(handler));               //** any path
    CHECK_RC(nng_http_handler_collect_body(handler, false, 0)); //** not collect body
    CHECK_RC(nng_http_handler_set_method(handler, NULL));       //** any method
    CHECK_RC(nng_http_server_add_handler(server, handler));
    if ((rc = nng_http_server_start(server)) != 0) {
        log_error("[%s]<nng_http_server> start http server failed: %s", "$KIND", nng_strerror(rc));
        return rc;
    }
    log_info("[%s]<nng_http_server> serve %s", "$KIND", url->u_rawurl);
    nng_url_free(url);
    nng_thread_destroy(hold);
    return rc;
}

```
